test/Mono.Linker.Tests.Cases/Advanced/TypeCheckRemoval.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Metadata;
+
+namespace Mono.Linker.Tests.Cases.Advanced
+{
+	[SetupCompileArgument ("/optimize+")]
+	class TypeCheckRemoval
+	{
+		public static void Main ()
+		{
+			TestTypeCheckRemoved_1 (null);
+			TestTypeCheckRemoved_2<string> (null);
+			TestTypeCheckRemoved_3 (null, null);
+
+			TestTypeCheckKept_1 ();
+			TestTypeCheckKept_2<string> (null);
+			TestTypeCheckKept_3 ();
+			TestTypeCheckKept_4 (null);
```

Comment:
I would also add one test which uses `as` - I know it translates to the same instruction, but it would be good to have it for completeness.